src/Tools/dotnet-monitor/CollectionRules/ActionOptionsDependencyAnalyzer.cs

```diff
@@ -0,0 +1,255 @@
+        }
+
+        private sealed class ActionDependency
+        {
+            public ActionDependency(CollectionRuleActionOptions action, string resultName)
+            {
+                Action = action;
+                ResultName = resultName;
+            }
+
+            public CollectionRuleActionOptions Action { get; }
+
+            public string ResultName { get; }
+
+            public string GetActionResultToken() => string.Concat(ActionReferencePrefix,
+                Action.Name, Separator, ResultName, SubstitutionSuffix);
+        }
+
+        public ActionOptionsDependencyAnalyzer(CollectionRuleContext context)
+        {
+            _ruleContext = context ?? throw new ArgumentNullException(nameof(context));
+        }
+
+        public IList<CollectionRuleActionOptions> GetActionDependencies(int actionIndex)
+        {
+            EnsureDependencies();
+
+            if (_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return properties
+                    .SelectMany(p => p.Value.ActionDependencies)
+                    .DistinctBy(a => a.Action.Name, StringComparer.OrdinalIgnoreCase)
+                    .Select(a => a.Action).ToArray();
+            }
+            return Array.Empty<CollectionRuleActionOptions>();
+        }
+
+        public object SubstituteOptionValues(int actionIndex, object settings)
+        {
+            EnsureDependencies();
+            if (!_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return settings;
+            }
+
+            if (settings is ICloneable cloneable)
+            {
+                settings = cloneable.Clone();
+            }
+            else
+            {
+                _ruleContext.Logger.InvalidSettings(settings.GetType().FullName);
+                return settings;
+            }
+
+            foreach (KeyValuePair<string, PropertyDependency> property in properties)
+            {
+                string newValue = property.Value.OriginalValue;
+                foreach(ActionDependency actionDependency in property.Value.ActionDependencies)
+                {
+                    string actionToken = actionDependency.GetActionResultToken();
+                    if (!_ruleContext.ActionResults.TryGetValue(actionDependency.Action.Name, out CollectionRuleActionResult results))
+                    {
+                        _ruleContext.Logger.InvalidResultReference(actionToken);
+                        continue;
+                    }
+                    if (!results.OutputValues.TryGetValue(actionDependency.ResultName, out string result))
+                    {
+                        _ruleContext.Logger.InvalidResultReference(actionToken);
+                        continue;
+                    }
+                    newValue = newValue.Replace(actionToken, result, StringComparison.OrdinalIgnoreCase);
```

Comment:
Because this is done iteratively, this could result in dynamic expansion in a limited sense so long as two or more action results are references initially and the value of the first action result contains the token of the subsequent action result. This is probably okay because its not recursive but could be exponential.

For example, if the following actions exist:
```json
[{
  "Name": "Action1",
  "Type": "ActionType1"
},{
  "Name": "Action2",
  "Type": "ActionType2"
},{
  "Type": "ActionType3",
  "Settings": {
    "SomeProperty": "$(Actions.Action1.Output1) $(Actions.Action2.Output1)"
  }
}]
```

If ActionType1 outputs a result called Output1 with a value of "$(Actions.Action2.Output1)" and ActionType2 outputs a result called Output1 with a value of "Hello", then the SomeProperty property of Action3 will have a value of "Hello Hello" passed into the action.

I don't think if this is something we need to prevent. If we do, we could replace all of the tokens in the property value with placeholders and then fill the placeholders individually rather than just doing the iterative find-and-replace.

Reply:
I decided to fix this by tracking each token location. That also allows me to do potentially less allocations on the final result.